File: src/j2ee/com.ibm.sbt.web/WebContent/js/sdk/sbt/base/BaseService.js
/*
 * © Copyright IBM Corp. 2012
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); 
 * you may not use this file except in compliance with the License. 
 * You may obtain a copy of the License at:
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software 
 * distributed under the License is distributed on an "AS IS" BASIS, 
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or 
 * implied. See the License for the specific language governing 
 * permissions and limitations under the License.
 */

/**
 * 
 * Javascript Base APIs for IBM Connections 
 * @module sbt.connections.BaseService
 * @author Carlos Manias
 * 
 */
define(['../declare','../config','../lang','../base/core','../xml','../xpath','../Cache','../Endpoint', '../base/BaseConstants', 
        "../validate", '../log', '../stringUtil', '../base/BaseHandler','../util','./BaseEntity'],
		function(declare,cfg,lang,con,xml,xpath,Cache,Endpoint, BaseConstants, validate, log, stringUtil, BaseHandler, util, BaseEntity) {
	
	var requests = {};
	
	function notifyCallbacks(id,param) {
		log.debug("notifyCallbacks() : called with id : {0}, and param : {1}", id, param);
		var r = requests[id];
  		if(r) {
	  		delete requests[id];
	  		for(var i=0; i<r.length; i++) {
	  			r[i](param);
	  		}
  		}
	}
	
	/**
	Base service class.

	@class BaseService
	@constructor
	@param {Object} options  Options object
	@param {String} [options.endpoint]  Endpoint to be used.
		
	**/
	var BaseService = declare(null, {
		"-chains-" : {
			constructor : "manual"
		},	
		
		_endpoint: null,
		
		constructor: function(_options) {			
			this._endpoint = Endpoint.find(_options.endpoint);
			this.Constants = lang.mixin(lang.mixin({}, BaseConstants), _options.Constants);
			this._con = _options.con || con; //NameSpaces
			var cacheSize = _options.cacheSize;
			if(cacheSize && cacheSize>0) {
				this._cache = new Cache(cacheSize);
			}			
		},
		
		_notifyResponse: function(args, response, summary){
			if (args.load || args.handle) {
				if (args.handle) {
					try {
						args.handle(response, summary);
					} catch (error) {
						log.error("Error running handle callback : {0}", error);
					}
				}
				if (args.load) {
					try {
						args.load(response, summary);
					} catch (error) {
						log.error("Error running load callback : {0}", error);
					}
				}
			}
		},
		
		_createEntityObject : function (service, id, requestArgs, args){
			return requestArgs.entity.apply(null,[service, id]);
		},
		
		_getOne: function(args,getArgs) {			
			var entity = this._createEntityObject(this, args.id, getArgs, args);			
			if(args.loadIt == false){
				this._notifyResponse(args,entity);
			}else{
				this._load(entity,args,getArgs);
			}
			return entity;
		},
		
		_load: function (entity,args,getArgs) {
			getArgs.methodType = "get";
			var _self = this;
			var handleAs = (getArgs.handleAs)?getArgs.handleAs : "text";
			var cachedData = null;
			if(getArgs.cachingEnabled){
				cachedData = this._find(entity._id);
			}
			if(cachedData) {				
					entity.setData(cachedData);		
					this._notifyResponse(args, entity);
			}else{
				if (getArgs.cachingEnabled && requests[entity._id]) {
					this._stackRequestsForExistingId(entity, args);				
				} else {
					if (getArgs.cachingEnabled) {
						this._stackRequestsForNewId(entity, args);
					}
					this._endpoint.xhrGet({
						serviceUrl:	_self._constructServiceUrl(getArgs),
						handleAs:	handleAs,
						content:	_self._constructQueryObj(args,getArgs),
						load: function(data, ioArgs) {
							_self._loadOnLoad(data, ioArgs, entity, args, getArgs);	
						},
						error: function(error){
							_self._loadOnError(error, args);
						}
					});
				}
			}
		},
		
		_loadOnLoad : function (data, ioArgs, entity, args, getArgs){
			var dataHandler = (getArgs.dataHandler) ? getArgs.dataHandler : entity._dataHandler ;			
			entity.setData(dataHandler._extractEntryFromSingleEntryFeed(data, ioArgs));			
			if (getArgs.cachingEnabled) {
				if(this._cache){
					this._cache.put(dataHandler._extractIdFromEntry(entity._data),entity._data);
				}
				notifyCallbacks(entity._id, entity);
			} else {
				this._notifyResponse(args, entity);
			}	
		},
		
		_loadOnError : function (error, args){
			util.notifyError(error, args);
		},
		
		_stackRequestsForExistingId : function (entity, args){
			if(args.load){
				requests[entity._id].push(args.load);
			}
			if(args.handle){
				requests[entity._id].push(args.handle);
			}
		},
		
		_stackRequestsForNewId : function (entity, args){
			if(args.load) {
				requests[entity._id] = [args.load];
			}
			if(args.handle){
				if(requests[entity._id]) {
					requests[entity._id].push(args.handle);
				}else {
					requests[entity._id] = [args.handle];
				}
			}	
		},
		
		getIdFromIdOrObject: function(inputEntity){
			return entityId = (!(typeof inputEntity == "object")) ? inputEntity._id : inputEntity;
		},
		
		_constructPayload: function(requestArgs){
			var dataHandler = (requestArgs.dataHandler) ? requestArgs.dataHandler : requestArgs.entity._dataHandler ;			
			return dataHandler._constructPayload(requestArgs);
		},
		
		_createRequestObject: function(args, requestArgs){
			var _self = this;
			var requestObject = {
					serviceUrl:_self._constructServiceUrl(requestArgs)									
			};
			if (args.parameters) {
				if(!(requestArgs.urlParams)){
					requestArgs.urlParams = {};
				}
				requestArgs.urlParams = lang.mixin(requestArgs.urlParams, args.parameters);
			}
			if(requestArgs.headers){
				requestObject.headers = requestArgs.headers;
			}
			if(requestArgs.handleAs){
				requestObject.handleAs = requestArgs.handleAs;
			}
			return requestObject;
		},
		
		/**
			Create a new entity
			
			@method createEntity
			@param {Object} entity  Entity object
			@param {Object} [args]  Argument object
			@param {Boolean} [args.loadIt=true] 
			@param {Function} [args.load] 
			@param {Function} [args.error] 
			@param {Function} [args.handle] 
		
		**/
		_createEntity: function (args, postArgs) {
			var _self = this;
			var entity = postArgs.entity;			
			var requestObject = lang.mixin({
				postData:_self._constructPayload(postArgs),
				load:function(data, ioArgs){
					_self._createEntityOnLoad(data, ioArgs, entity, args, postArgs);
				},
				error:function(error){
					_self._createEntityOnError(error, args);
				}
			}, _self._createRequestObject(args, postArgs));
			this._endpoint.xhrPost(requestObject);
		},
		
		_createEntityOnLoad : function (data, ioArgs, entity, args, postArgs){
			var dataHandler = (postArgs.dataHandler) ? postArgs.dataHandler : entity._dataHandler ;						
			entity._data = dataHandler._extractEntryFromSingleEntryFeed(data, ioArgs);
			this._notifyResponse(args,entity);
		},
		
		_createEntityOnError : function (error, args){
			util.notifyError(error, args);
		},
		
		/**
		Update an existing entity
		
		@method updateEntity
		@param {Object} entity  Entity object
		@param {Object} [args]  Argument object			
			@param {Function} [args.load] 
			@param {Function} [args.error] 
			@param {Function} [args.handle] 
		
		**/
		_updateEntity: function (args, putArgs) {
			
			var _self = this;
			var entity = putArgs.entity;			
			var requestObject = lang.mixin({
				putData:_self._constructPayload(putArgs),
				load:function(data, ioArgs){
					_self._updateEntityOnLoad(data, ioArgs, entity, args, putArgs);
				},
				error: function(error){
					_self._updateEntityOnError(error, args);
				}
			}, _self._createRequestObject(args, putArgs));
			this._endpoint.xhrPut(requestObject);
		},
		
		_updateEntityOnLoad : function (data, ioArgs, entity, args, putArgs){
			var dataHandler = (putArgs.dataHandler) ? putArgs.dataHandler : entity._dataHandler ;						
			entity._data = dataHandler._extractEntryFromSingleEntryFeed(data, ioArgs);
			this._notifyResponse(args,entity);
		},
		
		_updateEntityOnError : function (error, args){
			util.notifyError(error, args);
		},
			
		/**
		Delete an existing entity
		
		@method deleteEntity
		@param {String/Object} entity  id of the entity or the entity object.
		@param {Object} [args]  Argument object			
			@param {Function} [args.load] 
			@param {Function} [args.error] 
			@param {Function} [args.handle] 
		
		**/
		_deleteEntity: function (args, deleteArgs) {
			var _self = this;
			var entity = deleteArgs.entity;			
			var requestObject = lang.mixin({
				load:function(data, ioArgs){
					_self._deleteEntityOnLoad(data, ioArgs, entity, args);
				},
				error: function(error){
					_self._deleteEntityOnError(error, args);
				}
			}, _self._createRequestObject(args, deleteArgs));
			this._endpoint.xhrDelete(requestObject);
		},
		
		_deleteEntityOnLoad : function (data, ioArgs, entity, args){
			if(this._cache) {
				this._deleteIdFromCache(entity._id);
      		}
			this._notifyResponse(args);
		},
		
		_deleteEntityOnError : function (error, args){
			util.notifyError(error, args);
		},
		
		_constructServiceUrl: function (requestArgs){			
			var authType = requestArgs.authType;
			authType = (requestArgs.authType)?requestArgs.authType : "";
			
			var url = this.Constants.entityServiceBaseUrl + authType + (this.Constants.serviceEntity[requestArgs.serviceEntity]?this.Constants.serviceEntity[requestArgs.serviceEntity]: "") + (this.Constants.entityType[requestArgs.entityType] ? this.Constants.entityType[requestArgs.entityType] : "");
			
			if (requestArgs.replaceArgs) {
				url = stringUtil.replace(url, requestArgs.replaceArgs);
			}
			
			if(requestArgs.methodType != "get"){
				if (requestArgs.urlParams) {
					url += "?";
					var additionalElement = false;
					for (param in requestArgs.urlParams) {
						if (additionalElement) {
							url+= "&";
						}
						else {
							additionalElement = true;
						}// encode url parameters
						url += param + "=" + encodeURIComponent(requestArgs.urlParams[param]);
					}
				}
			}
			return url;
		},
		
		_constructQueryObj: function (args, requestArgs){
			var params = args.parameters;
			if(requestArgs.urlParams){				
				params = lang.mixin(params, requestArgs.urlParams);
			}
			return params;
		},
		
		_getEntities: function (args,getArgs) {
			var _self=this;
			getArgs.methodType = "get";	
			var requestObject = lang.mixin({
				content: _self._constructQueryObj(args,getArgs),
				load:function(data, ioArgs){
					summary = _self._getSummaryOnLoad(data, getArgs, args);
					entities = _self._getEntititiesOnLoad(data, getArgs, args);
					_self._notifyResponse(args, entities, summary);
				},
				error: function(error){
					_self._getEntitiesOnError(error, args);
				}
				
			}, _self._createRequestObject(args, getArgs));			
			this._endpoint.xhrGet(requestObject);
		},
		
		_getEntitiesOnError : function (error, args){
			util.notifyError(error, args);
		},
		
		_getSummaryOnLoad: function (data, getArgs, args){
			var dataHandler = getArgs.dataHandler;			
			return dataHandler._extractSummaryFromEntitiesFeed(data);
		},
		
		_getEntititiesOnLoad: function (data, getArgs, args){			
			var dataHandler = getArgs.dataHandler ;			
			var entities = [];			
			var entry = dataHandler._extractEntriesFromEntitiesFeed(data);
			for(var count=0; count < entry.length; count++){
				 var node = entry[count];
				 var entityId = dataHandler._extractIdFromEntry(node);
				 entityId = getArgs.parseId ? getArgs.parseId(entityId) : entityId;			
				 var entity = this._createEntityObject(this, entityId, getArgs, args);
				if(!(getArgs.ignoreData)){
					entity.setData(node);
				}
				entities.push(entity);
			}
			return entities;
		},
		
		_isEmail: function(id) {
			return id && id.indexOf('@')>=0;
		},
		_deleteIdFromCache:function(_id){
			if(this._cache) {
				this._cache.remove(_id);
			}
		},		
		_find: function(id) {
			if(this._cache) {
				return this._cache.get(id);
			}
			else{			
				return null;
			}
		}
	});
	return BaseService;
});

File: sdk/com.ibm.sbt.core/src/main/java/com/ibm/sbt/services/client/base/BaseService.java
/*
 * © Copyright IBM Corp. 2012
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.ibm.sbt.services.client.base;

import static com.ibm.sbt.services.client.base.CommonConstants.APPLICATION_ATOM_XML;
import static com.ibm.sbt.services.client.base.CommonConstants.AT;
import static com.ibm.sbt.services.client.base.CommonConstants.CONTENT_TYPE;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;

import com.ibm.commons.util.StringUtil;
import com.ibm.sbt.services.client.ClientService;
import com.ibm.sbt.services.client.ClientService.Handler;
import com.ibm.sbt.services.client.ClientServicesException;
import com.ibm.sbt.services.client.Response;
import com.ibm.sbt.services.client.base.CommonConstants.HTTPCode;
import com.ibm.sbt.services.client.base.datahandlers.EntityList;
import com.ibm.sbt.services.endpoints.Endpoint;
import com.ibm.sbt.services.endpoints.EndpointFactory;

/**
 * This class defines common behaviour for all the services 
 * @author Carlos Manias
 *
 */
public abstract class BaseService implements Serializable {

	private static final long 					serialVersionUID = 1L;

	public static final int						DEFAULT_CACHE_SIZE		= 0;
	public static final String					DEFAULT_ENDPOINT_NAME	= "connections";
	
	protected transient static HashMap<String, Object>	cache					= new HashMap<String, Object>();
	protected transient int						cacheSize;
	protected Handler							dataFormat;
	protected transient Endpoint				endpoint;

	protected String[] serviceMappingKeys = new String[]{};
	/**
	 * Constructor
	 */
	public BaseService() {
		this(DEFAULT_ENDPOINT_NAME, DEFAULT_CACHE_SIZE);
	}

	/**
	 * Constructor
	 * 
	 * @param endpointName
	 */
	public BaseService(String endpointName) {
		this(endpointName, DEFAULT_CACHE_SIZE);
	}

	/**
     * Constructor
     * 
     * @param endpointName
     * @param cacheSize
     */
    public BaseService(String endpointName, int cacheSize) {
        if (StringUtil.isEmpty(endpointName)) {
            endpointName = DEFAULT_ENDPOINT_NAME;
        }
        
        this.endpoint = EndpointFactory.getEndpointFromEnvironment(endpointName, null);
        this.cacheSize = cacheSize;
    }

	/**
     * Constructor
     * 
     * @param endpoint
     * @param cacheSize
     */
    public BaseService(Endpoint endpoint) {
        this(endpoint, DEFAULT_CACHE_SIZE);
    }

	/**
     * Constructor
     * 
     * @param endpoint
     * @param cacheSize
     */
    public BaseService(Endpoint endpoint, int cacheSize) {
        this.endpoint = endpoint;
        this.cacheSize = cacheSize;
    }

	/**
	 * @return dataFormat
	 */
	public Handler getDataFormat() {
		return this.dataFormat;
	}

	/**
	 * Sets the endpoint
	 * 
	 * @param endpoint
	 *            the endpoint
	 */
	public void setEndpoint(Endpoint endpoint) {
		this.endpoint = endpoint;
	}

	/**
	 * Returns the Endpoint
	 * 
	 * @return endpoint the Endpoint
	 */
	public Endpoint getEndpoint() {
		return endpoint;
	}

    /**
     * Get authentication type for the endpoint. like basicAuth, oauth etc.
     * @return
     */
	public NamedUrlPart getAuthType(){
		return new NamedUrlPart("authType",AuthType.getAuthTypePart(endpoint));
    }
	
	/**
	 * Returns a Version object with the API version of the Endpoint
	 * @return
	 */
	public Version getApiVersion(){
		return Version.parse(endpoint.getApiVersion());
	}
	
	/**
	 * Subclasses must list the context roots of the Connections APIs they use.
	 * @return
	 */
	public String[] getServiceMappingKeys(){
		return serviceMappingKeys;
	}
	
	/**
	 * Returns either the configured or the default serviceMappings for the service 
	 * @return
	 */
	public NamedUrlPart[] getServiceMappings(){
		String[] serviceMappingKeys = getServiceMappingKeys();
		Map<String, String> serviceMappings = getEndpoint().getServiceMappings();
		
		NamedUrlPart[] urlParts = new NamedUrlPart[serviceMappingKeys.length];
		for(int i = 0;i < serviceMappingKeys.length; i++){
			String key = serviceMappingKeys[i];
			if (serviceMappings.containsKey(key)){
				urlParts[i] = new NamedUrlPart(key, serviceMappings.get(key));
			}else{
				urlParts[i] = new NamedUrlPart(key, key);
			}
		}
		
		return urlParts;
	}
    
	/**
	 * Return the size of the cache
	 * 
	 * @return cacheSize the cache size
	 */
	public int getCacheSize() {
		return cacheSize;
	}

	/**
	 * Sets the size of the cache
	 * 
	 * @param cacheSize
	 */
	public void setCacheSize(int cacheSize) {
		this.cacheSize = cacheSize;
	}

	/*
	 * This method makes a network call and returns an entity
	 */
	protected <T extends BaseEntity> T getEntity(String url, Map<String, String> parameters, IFeedHandler<T> feedHandler) throws ClientServicesException {
		try {
			Response response = retrieveData(url, parameters);
			return feedHandler.createEntity(response);
		} catch (Exception e){
			throw new ClientServicesException(e);
		}
	}
	
	/*
	 * This method makes a network call and returns a Collection of Entities
	 */
	protected <T extends BaseEntity> EntityList<T> getEntities(String url, Map<String, String> parameters, IFeedHandler<T> feedHandler) throws ClientServicesException {
		try {
			Response dataHolder = retrieveData(url, parameters);
			return feedHandler.createEntityList(dataHolder);
		} catch (Exception e){
			throw new ClientServicesException(e);
		}
	}

	/*
     * This method makes a network call and returns a Collection of Entities
     */
    protected <T extends BaseEntity> EntityList<T> getEntities(String url, Map<String, String> parameters, Map<String, String> headers, IFeedHandler<T> feedHandler) throws ClientServicesException {
		try {
	        Response dataHolder = retrieveData(url, parameters, headers, null);
	        return feedHandler.createEntityList(dataHolder);
		} catch (Exception e){
			throw new ClientServicesException(e);
		}
    }


	/**
	 * Post some data and return the resulting response
	 * 
	 * @param serviceUrl
	 * @param parameters
	 * @param headers
	 * @param content
	 * @param DateFormat
	 * @return
	 * @throws ClientServicesException
	 */
    public Response createData(String serviceUrl, Map<String, String> parameters, Map<String,String> headers, Object content) throws ClientServicesException {
        return createData(serviceUrl, parameters, headers, content, getDataFormat());
    }
	
    /**
     * Post some data and return the resulting response
     * 
     * @param serviceUrl
     * @param parameters
     * @param headers
     * @param content
     * @param format
     * @return
     * @throws ClientServicesException
     */
    public Response createData(String serviceUrl, Map<String, String> parameters, Map<String,String> headers, Object content, Handler format) throws ClientServicesException {
        Response result = getClientService().post(serviceUrl, parameters, headers, content, format);
        return result;
    }
    	
	/**
	 * Creates data and returns the result
	 * 
	 * @param clientService
	 * @param serviceUrl
	 * @param parameters
	 * @param content
	 * @return the result of the creation operations
	 * @throws ClientServicesException
	 *             when the creation fails, as null/false return may have other meaning here
	 */
	public Response createData(String serviceUrl, Map<String, String> parameters, Object content, Handler format) throws ClientServicesException {
    	return createData(serviceUrl, parameters, new HashMap<String, String>(), content);	    
	}
	
	
	/**
	 * Creates data and returns the result 
	 * @param serviceUrl
	 * @param parameters
	 * @param content
	 * @return
	 * @throws ClientServicesException
	 */
	public Response createData(String serviceUrl, Map<String, String> parameters, Object content) throws ClientServicesException {
		return createData(serviceUrl, parameters, content, getDataFormat());
	}

	/**
	 * This method encapsulate the access to the Endpoint client service.
	 * 
	 * @return the client serivce associated with the current Endpoint
	 * @throws ClientServicesException
	 */
	protected ClientService getClientService() throws ClientServicesException {
		return endpoint.getClientService();
	}

	
	/**
	 * Returns true if delete completed successfully and false if an error happened
	 * 
	 * @param clientService
	 * @param serviceUrl
	 * @param parameters
	 * @param nameParameterId
	 * @return
	 * @throws ClientServicesException
	 */
	public Response deleteData(String serviceUrl, Map<String, String> parameters, String nameParameterId)
			throws ClientServicesException {
	    return deleteData(serviceUrl, parameters, new HashMap<String, String>(), nameParameterId);
	}
	
	/**
	 * execute a delete and returns the result
	 * @param serviceUrl
	 * @param parameters
	 * @param headers
	 * @param nameParameterId
	 * @return
	 * @throws ClientServicesException
	 */
    public Response deleteData(String serviceUrl, Map<String, String> parameters, Map<String, String> headers, String nameParameterId)
            throws ClientServicesException {
        String uniqueId = null;
        if (nameParameterId != null) {
            uniqueId = (parameters == null) ? nameParameterId : parameters.get(nameParameterId);
        }

        Response r = getClientService().delete(serviceUrl, parameters, headers, getDataFormat());
        
        if (cacheSize > 0 && uniqueId != null) {
            removeFromCache(uniqueId);
        }

        return r;
    }
    
    public Response delete(String serviceUrl, Map<String, String> parameters, Map<String, String> headers, String nameParameterId, String content)throws ClientServicesException{
    	String uniqueId = null;
        if (nameParameterId != null) {
            uniqueId = (parameters == null) ? nameParameterId : parameters.get(nameParameterId);
        }

        Response r = getClientService().delete(serviceUrl, parameters, headers, getDataFormat(),content);
        
        if (cacheSize > 0 && uniqueId != null) {
            removeFromCache(uniqueId);
        }

        return r;
    }
	/**
	 * retrieveData
	 * 
	 * @param clientService
	 * @param url
	 * @param parameters
	 * @return Convenience method for retrieving multiple JsonObjects from the server
	 * @throws ClientServicesException
	 */
	public Response retrieveData(String url, Map<String, String> parameters) throws ClientServicesException {
		return retrieveData(url, parameters, null);
	}

	/**
	 * execute a get and return the result
	 * 
	 * @param url
	 * @param parameters
	 * @param headers
	 * @param nameParameterId
	 * @return
	 * @throws ClientServicesException
	 */
	public Response retrieveData(String url, Map<String, String> parameters, Map<String, String> headers, String nameParameterId)
	            throws ClientServicesException {
	        Object data = null;
	        Response dataHolder = null;
	        String uniqueId = "";
	        if (nameParameterId != null) {
	            uniqueId = parameters.get(nameParameterId);
	            data = findInCache(uniqueId);
	            dataHolder = new Response(data);
	        }
	        if (data == null) {
	            dataHolder = getClientService().get(url, parameters, headers, getDataFormat());
	            //in case of 401 errors  client service returns null
	            if (dataHolder == null) return null;
	            data = dataHolder.getData();
	            if (cacheSize > 0 && nameParameterId != null) {
	                addDataToCache(uniqueId, data);
	            }
	        }
	        return dataHolder;
	    }
	
	/**
	 * retrieveData()
	 * 
	 * @param clientService
	 * @param url
	 * @param parameters
	 * @param nameParameterId
	 * @return Method to retrieve a single Object from the server
	 * @throws ClientServicesException
	 */
	//TODO Fix cache with DataHolder object
	public Response retrieveData(String url, Map<String, String> parameters, String nameParameterId) throws ClientServicesException {
			return retrieveData(url, parameters, new HashMap<String, String>(), nameParameterId);
	}
	
	/**
	 * Returns true if update completed successfully and false if an error happened
	 * 
	 * @param clientService
	 * @param serviceUrl
	 * @param parameters
	 * @param content
	 * @param nameParameterId - only used to distinguish between cache entries, 
	 *     must be a valid key of the parameters map
	 * @return
	 * @throws ClientServicesException
	 */
	//TODO: Use Args pattern to pass the headers and avoid hard-coding the content-type
	public Response updateData(String serviceUrl, Map<String, String> parameters, Object content,
			String nameParameterId) throws ClientServicesException {
	    return updateData(serviceUrl, parameters, new HashMap<String, String>(), content, nameParameterId);
	}

	/**
	 * execute a put and return the response
	 * 
	 * @param serviceUrl
	 * @param parameters
	 * @param headers
	 * @param content
	 * @param nameParameterId
	 * @return
	 * @throws ClientServicesException
	 */
   public Response updateData(String serviceUrl, Map<String, String> parameters, Map<String, String> headers, Object content, String nameParameterId) throws ClientServicesException {
        String uniqueId = "";
        if (nameParameterId != null) {
            uniqueId = parameters.get(nameParameterId);
        }

        if (!headers.containsKey(CONTENT_TYPE)) {
            // TODO shouldn't assume this
        	headers.put(CONTENT_TYPE, APPLICATION_ATOM_XML);
        }
        Response result = getClientService().put(serviceUrl, parameters, headers, content, getDataFormat());
        if (cacheSize > 0 && nameParameterId != null) {
            addDataToCache(uniqueId, content);
        }

        return result;
   }

   public Response putData(String serviceUrl, Map<String, String> parameters, Map<String, String> headers, Object content, String uniqueId) throws ClientServicesException {
       Response result = getClientService().put(serviceUrl, parameters, headers, content, getDataFormat());
       
       if (cacheSize > 0 && uniqueId != null) {
           addDataToCache(uniqueId, content);
       }

       return result;
   }
   
   protected void checkResponseCode(Response response, HTTPCode expectedCode) throws ClientServicesException {
	   if (response != null && response.getResponse() != null && response.getRequest()!=null) {
		   if (response.getResponse().getStatusLine() != null && expectedCode.checkCode(response.getResponse().getStatusLine().getStatusCode())) {
			   return;
		   } else {
			   throw new ClientServicesException(response.getResponse(), response.getRequest());
		   }
	   } else {
		   throw new ClientServicesException(null, "Response is null");
	   }
   }
	
	/**
	 * findInCache()
	 * 
	 * @param userId
	 * @return Method to search the cache
	 */
	private Object findInCache(String key) {

		Object data = null;

		if (cache.containsKey(key)) {
			data = cache.get(key);
		} else {
			// Cache miss
		}

		return data;
	}

	/**
	 * Removes from cache the data stored under the given key
	 * 
	 * @param key
	 */
	protected void removeFromCache(String key) {
		cache.remove(key);
	}

	/**
	 * Adds to cache the object using the given key
	 * 
	 * @param key
	 * @param content
	 */
	protected void addDataToCache(String key, Object content) {
		// Limit the cache size as per options
		// to check if cache is full , remove if full using LRU algorithm
		cache.put(key, content);
	}

	
	/**
	 * Returns a valid parameters HashMap even if null is passed
	 * 
	 * @param parameters
	 * @return
	 */
	protected Map<String, String> getParameters(Map<String, String> parameters) {
		if(parameters == null) return new HashMap<String, String>();
		else return parameters;
	}
	
	/**
	 * Return true if the id as email address
	 * 
	 * @param id
	 * @return
	 */
	protected boolean isEmail(String id) {
		return (id == null) ? false : id.contains(AT);
	}
	
}
